### PR TITLE
HPCC-25755 Add general file/folder watcher mechanism

### PR DESCRIPTION
**Pull request by jakesmith**
Signed-off-by: Jake Smith <jake.smith@lexisnexisrisk.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

File: system/jlib/jfile.hpp [hunk 1]
```diff
@@ -667,4 +671,56 @@ inline bool canRename(const char * filename) { return queryFileSystemProperties(
...
+    opened        = 0x020,
+    movedFrom     = 0x040,
+    movedTo       = 0x080,
+    created       = 0x100,
+    deleted       = 0x200,
+#endif
+    anyChange     = created | movedTo | movedFrom | closedWrite | modified | metaModified
+};
+
+inline FileWatchEvents operator|(FileWatchEvents a, FileWatchEvents b)
```
Comment (richardkchapman) [inline system/jlib/jfile.hpp:705]: Should this use BITMASK_ENUM ?

File: system/jlib/jfile.cpp [hunk 2]
```diff
@@ -19,6 +19,9 @@
 #ifdef _WIN32
 #include <errno.h>
 //#include <winsock.h>  // for TransmitFile
+#else
+#include <poll.h>
+#include <sys/inotify.h>
```
Comment (richardkchapman) [inline system/jlib/jfile.cpp:24]: This does not compile on OSX

File: system/jlib/jfile.cpp [hunk 3]
```diff
@@ -7106,3 +7109,230 @@ const FileSystemProperties & queryFileSystemProperties(const char * filename)
...
+        }
+        std::string filename;
+        int watchFd = -1;
+        FileWatchEvents events = FileWatchEvents::none;
+        bool pendingRemoval = false;
+    };
+    int inotifyFd = -1;
+    int pipefd[2] = {-1, -1};
+    pollfd toPoll[2];
+    std::vector<Owned<CMonitorItem>> monitoredFiles;
```
Comment (richardkchapman) [inline system/jlib/jfile.cpp:7131]: Is this what we are doing instead of using jarray these days?

File: system/jlib/jfile.cpp [hunk 4]
```diff
@@ -7106,3 +7109,230 @@ const FileSystemProperties & queryFileSystemProperties(const char * filename)
...
+
+            do
+            {
+                std::string filename;
+                if (event->len) // if watching directory event->name is populated with file within that dir.
+                    filename = event->name;
+                else
+                {
+                    /* if watching a file, lookup filename from event handle
+                     * linear search, not worth hash table, not expected to be that many.
```
Comment (richardkchapman) [inline system/jlib/jfile.cpp:7161]: Famous last words... is a hash table really that hard?

File: system/jlib/jfile.cpp [hunk 5]
```diff
@@ -7106,3 +7109,230 @@ const FileSystemProperties & queryFileSystemProperties(const char * filename)
...
+        {
+            Owned<IException> e = makeErrnoException("CFileEventWatcher::stop() - write -> pipefd[1]");
+            EXCLOG(e);
+        }
+        threaded.join();
+    }
+    virtual bool add(const char *filename, FileWatchEvents events) override
+    {
+        CriticalBlock b(crit);
+        for (const auto &item: monitoredFiles)
```
Comment (richardkchapman) [inline system/jlib/jfile.cpp:7249]: Hash table?

File: system/jlib/jfile.cpp [hunk 6]
```diff
@@ -7106,3 +7109,230 @@ const FileSystemProperties & queryFileSystemProperties(const char * filename)
...
+#ifndef _WIN32
+class CFileEventWatcher : public CInterfaceOf<IFileEventWatcher>, implements IThreaded
+{
+    class CMonitorItem : public CInterface
+    {
+    public:
+        CMonitorItem(const char *_filename, int _watchFd, FileWatchEvents _events) : filename(_filename), watchFd(_watchFd), events(_events)
+        {
+        }
+        std::string filename;
```
Comment (richardkchapman) [inline system/jlib/jfile.cpp:7123]: StringAttr might be better - I assume immutable

**Review by richardkchapman — CHANGES_REQUESTED** [review]
Mostly style questions, but need to disable (or implement) on OSX before can be merged. For current anticipated uses, disabling is fine.

File: system/jlib/jfile.hpp (hunk 1 above)
Comment (jakesmith) [reply system/jlib/jfile.hpp]: yes, I thought something like that had been introduced, but couldn't find it..
Have switched to it, also amended it to add operator &

File: system/jlib/jfile.cpp (hunk 2 above)
Comment (jakesmith) [reply system/jlib/jfile.cpp]: have moved into __linux__

File: system/jlib/jfile.cpp (hunk 3 above)
Comment (jakesmith) [reply system/jlib/jfile.cpp]: yes in general, for an owned arrays, and maps, tending to use stl.
More flexible, easier to use, and at least for maps, e.g. unordered_map, a fair bit faster if I remember correctly, although not massively time critical in this context.

File: system/jlib/jfile.cpp (hunk 4 above)
Comment (jakesmith) [reply system/jlib/jfile.cpp]: have converted to use ht(s)

File: system/jlib/jfile.cpp (hunk 5 above)
Comment (jakesmith) [reply system/jlib/jfile.cpp]: need 2 (lookup by fd and filename), have changed to use ht's.

File: system/jlib/jfile.cpp (hunk 6 above)
Comment (jakesmith) [reply system/jlib/jfile.cpp:7125]: std::string easier to use with other stl containers.

File: system/jlib/jfile.hpp [hunk 7]
```diff
@@ -667,4 +671,54 @@ inline bool canRename(const char * filename) { return queryFileSystemProperties(
...
+    closed        = 0x004,
+    closedWrite   = 0x008,
+    metaModified  = 0x010,
+    opened        = 0x020,
+    movedFrom     = 0x040,
+    movedTo       = 0x080,
+    created       = 0x100,
+    deleted       = 0x200,
+#endif
+    anyChange     = created | movedTo | movedFrom | closedWrite | modified | metaModified
```
Comment (ghalliday) [inline system/jlib/jfile.hpp:702]: and deleted?

File: system/jlib/jfile.cpp [hunk 8]
```diff
@@ -7106,3 +7111,244 @@ const FileSystemProperties & queryFileSystemProperties(const char * filename)
...
+            }
+            flags = fcntl(pipefd[i], F_GETFD, 0);
+            if (-1 != flags)
+            {
+                flags |= FD_CLOEXEC;
+                fcntl(pipefd[i], F_SETFD, flags);
+            }
+        }
+
+        toPoll[0].fd = pipefd[0];
```
Comment (ghalliday) [inline system/jlib/jfile.cpp:7214]: Could do with some explanation about what the pipe is for.  I think it is to allow the watcher to be cancelled, but not at all clear.

File: system/jlib/jfile.cpp [hunk 9]
```diff
@@ -7106,3 +7111,244 @@ const FileSystemProperties & queryFileSystemProperties(const char * filename)
...
+    int inotifyFd = -1;
+    int pipefd[2] = {-1, -1};
+    pollfd toPoll[2];
+    std::unordered_map<int, Linked<CMonitoredItem>> monitoredFilesByFd;
+    std::unordered_map<std::string, Linked<CMonitoredItem>> monitoredFiles;
+    std::unordered_set<int> pendingRemovals;
+    CriticalSection crit;
+    FileWatchFunc callback;
+    CThreaded threaded;
+    bool stopped = true;
```
Comment (ghalliday) [inline system/jlib/jfile.cpp:7139]: should be an atomic

File: system/jlib/jfile.cpp [hunk 10]
```diff
@@ -7106,3 +7111,244 @@ const FileSystemProperties & queryFileSystemProperties(const char * filename)
...
+                            {
+                                CMonitoredItem *item = iter->second;
+                                if (item->pendingRemoval) // else fd has been reused/readded
+                                {
+                                    monitoredFiles.erase(item->filename);
+                                    monitoredFilesByFd.erase(iter);
+                                }
+                            }
+                        }
+                    }
```
Comment (ghalliday) [inline system/jlib/jfile.cpp:7331]: Should clear pendingRemovals (outside of cs)

File: system/jlib/jfile.cpp [hunk 11]
```diff
@@ -7106,3 +7111,244 @@ const FileSystemProperties & queryFileSystemProperties(const char * filename)
...
+    }
+    void checkClose(int h)
+    {
+        if (-1 != h)
+            close(h);
+    }
+public:
+    CFileEventWatcher(FileWatchFunc _callback) : callback(_callback), threaded("CFileEventWatcher", this)
+    {
+        inotifyFd = inotify_init1(IN_NONBLOCK);
```
Comment (ghalliday) [inline system/jlib/jfile.cpp:7193]: Should this also set IN_CLOEXEC?

File: system/jlib/jfile.cpp [hunk 12]
```diff
@@ -7106,3 +7111,244 @@ const FileSystemProperties & queryFileSystemProperties(const char * filename)
...
+
+                        // clear any pending deletes
+                        CriticalBlock b(crit);
+                        for (const auto fd: pendingRemovals)
+                        {
+                            auto iter = monitoredFilesByFd.find(fd);
+                            if (iter != monitoredFilesByFd.end())
+                            {
+                                CMonitoredItem *item = iter->second;
+                                if (item->pendingRemoval) // else fd has been reused/readded
```
Comment (ghalliday) [inline system/jlib/jfile.cpp:7324]: I don't think this is ever cleared, so I'm not sure the check is required - comment is confusing...

**Review by ghalliday — COMMENTED** [review]
@jakesmith generally looks good.  A few comments.
It seems to overlap with the monitorDirectory() functionality - which seems to be barely used, but probably also works for windows.

File: system/jlib/jfile.cpp (hunk 8 above)
Comment (jakesmith) [reply system/jlib/jfile.cpp:7214]: yes, agree, I'll comment to explain.

File: system/jlib/jfile.hpp (hunk 7 above)
Comment (jakesmith) [reply system/jlib/jfile.hpp:702]: yes, I've added

File: system/jlib/jfile.cpp (hunk 9 above)
Comment (jakesmith) [reply system/jlib/jfile.cpp:7139]: will change.

File: system/jlib/jfile.cpp (hunk 10 above)
Comment (jakesmith) [reply system/jlib/jfile.cpp:7331]: I forgot to clear them.. I'll change,
but I need to do within the cs, others could be adding/removing and updating pendingRemovals on other threads.

File: system/jlib/jfile.cpp (hunk 11 above)
Comment (jakesmith) [reply system/jlib/jfile.cpp:7193]: yes think it should, so should the pipe handles, and I see that can use pipe2 to create those without the fcntl code.
Will change.

File: system/jlib/jfile.cpp (hunk 12 above)
Comment (jakesmith) [reply system/jlib/jfile.cpp:7324]: The item is deleted (last ownership by one of the iterators), but pendingRemovals should be clearered too, fixing.

Comment (jakesmith) [reply system/jlib/jfile.cpp:7324]: will clarify comment

File: system/jlib/jfile.cpp (hunk 10 above)
Comment (ghalliday) [reply system/jlib/jfile.cpp:7325]: good point

File: system/jlib/jfile.cpp [hunk 13]
```diff
@@ -7106,3 +7112,237 @@ const FileSystemProperties & queryFileSystemProperties(const char * filename)
...
+            close(h);
+    }
+public:
+    CFileEventWatcher(FileWatchFunc _callback) : callback(_callback), threaded("CFileEventWatcher", this)
+    {
+        inotifyFd = inotify_init1(IN_NONBLOCK|IN_CLOEXEC);
+        if (-1 == inotifyFd)
+            throw makeStringException(-1, "CFileEventWatcher - inotify_init1");
+
+
```
Comment (ghalliday) [inline system/jlib/jfile.cpp:7198]: trivial: extra newline

**Review by ghalliday — APPROVED** [review]
Looks good to me - please squash

File: system/jlib/jfile.cpp (hunk 13 above)
Comment (jakesmith) [reply system/jlib/jfile.cpp]: removed

**Comment by jakesmith** [comment]
@ghalliday - squashed.

**Review by richardkchapman — APPROVED** [review]

**Pull request merged**